package/kedro_viz/integrations/kedro/run_hooks.py

```diff
@@ -0,0 +1,249 @@
+"""`kedro_viz.integrations.kedro.run_hooks` defines hooks to add additional functionalities for a kedro run."""
+
+import logging
+import traceback
+from time import perf_counter
+from typing import Any, Optional, Union
+
+from kedro.framework.hooks import hook_impl
+from kedro.pipeline.node import Node as KedroNode
+
+from kedro_viz.integrations.kedro.hooks_utils import (
+    create_dataset_event,
+    generate_timestamp,
+    hash_node,
+    is_default_run,
+    write_events,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class PipelineRunStatusHook:
+    """
+    Collect and write pipeline and dataset events during Kedro runs.
+
+    This hook class tracks pipeline execution events. It integrates with Kedro's hook
+    system to monitor pipeline runs and persist execution data for visualization.
+
+    Attributes:
+        _node_start (dict[str, float]): Mapping of node names to their start times
+            for performance tracking.
+        _events (list[dict[str, Any]]): List of all events collected during the run.
+        _datasets (dict[str, Any]): Dictionary of dataset configurations from the
+            Kedro data catalog for size and metadata lookups.
+        _current_node (Optional[KedroNode]): Currently executing node, used for
+            error context and tracking.
+        _current_dataset (Optional[str]): Currently processed dataset name,
+            used for error context during I/O operations.
+        _current_operation (Optional[str]): Current dataset operation type
+            (e.g., 'loading', 'saving'), used for error context.
+        _all_nodes (list[KedroNode]): Complete list of nodes in the pipeline,
+            used for tracking execution progress.
+        _started_nodes (set[str]): Set of node names that have begun execution,
+            used to identify unstarted nodes during pipeline errors.
+    """
+
+    def __init__(self):
+        """
+        Initialize the pipeline run status hook.
+
+        Sets up empty tracking structures for monitoring pipeline execution,
+        including event collection, timing data, and error context management.
+        """
+        # Track times, events, and context for errors
+        self._node_start: dict[str, float] = {}
+        self._events: list[dict[str, Any]] = []
+        self._datasets: dict[str, Any] = {}
+        self._current_node: Optional[Any] = None
+        self._current_dataset: Optional[str] = None
+        self._current_operation: Optional[str] = None
+        self._all_nodes: list[KedroNode] = []
+        self._started_nodes: set[str] = set()
+
+    def _write_events(self) -> None:
+        """Persist events list to the project's .viz JSON file."""
+        write_events(self._events)
```

Comment:
What's the reasoning here with private and public methods? It seems odd to me to have a private and public method with the same name where one just calls the other. I'd suggest to just move the `write_events` method to here and not have the private wrapper.